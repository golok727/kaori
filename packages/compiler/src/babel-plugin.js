import * as t from "@babel/types";
const KAORI_PACKAGE_NAME = "kaori.js";
function findNonConflictingName(baseName, existingNames) {
    if (!existingNames.has(baseName)) {
        return baseName;
    }
    let counter = 1;
    let candidate = `${baseName}${counter}`;
    while (existingNames.has(candidate)) {
        counter++;
        candidate = `${baseName}${counter}`;
    }
    return candidate;
}
function jsxToLitHtmlPlugin() {
    return {
        name: "jsx-to-lit-html",
        visitor: {
            Program: {
                enter(path, state) {
                    state.needsComponentImport = false;
                    state.hasComponentImport = false;
                    state.needsHtmlImport = false;
                    state.hasHtmlImport = false;
                    // First, check what's already imported from kaori and their local names
                    const kaoriImports = new Map(); // imported name -> local name
                    path.traverse({
                        ImportDeclaration(innerPath) {
                            if (t.isStringLiteral(innerPath.node.source) &&
                                (innerPath.node.source.value === KAORI_PACKAGE_NAME ||
                                    innerPath.node.source.value.includes("kaori"))) {
                                innerPath.node.specifiers.forEach((spec) => {
                                    if (t.isImportSpecifier(spec) &&
                                        t.isIdentifier(spec.imported)) {
                                        const importedName = spec.imported.name;
                                        const localName = spec.local.name;
                                        kaoriImports.set(importedName, localName);
                                    }
                                });
                            }
                        },
                    });
                    // Use existing imports if available, otherwise find non-conflicting names
                    if (kaoriImports.has("html")) {
                        state.htmlTagName = kaoriImports.get("html");
                        state.hasHtmlImport = true;
                    }
                    if (kaoriImports.has("component")) {
                        state.componentFunctionName = kaoriImports.get("component");
                        state.hasComponentImport = true;
                    }
                    // Only scan for conflicts if we need to create new names
                    if (!state.hasHtmlImport || !state.hasComponentImport) {
                        const existingNames = new Set();
                        path.traverse({
                            Identifier(innerPath) {
                                existingNames.add(innerPath.node.name);
                            },
                        });
                        if (!state.hasHtmlImport) {
                            state.htmlTagName = findNonConflictingName("html", existingNames);
                        }
                        if (!state.hasComponentImport) {
                            state.componentFunctionName = findNonConflictingName("component", existingNames);
                        }
                    }
                },
                exit(path, state) {
                    // Add imports if needed
                    const imports = [];
                    if (state.needsComponentImport && !state.hasComponentImport) {
                        imports.push(t.importSpecifier(t.identifier(state.componentFunctionName), t.identifier("component")));
                    }
                    if (state.needsHtmlImport && !state.hasHtmlImport) {
                        imports.push(t.importSpecifier(t.identifier(state.htmlTagName), t.identifier("html")));
                    }
                    if (imports.length > 0) {
                        const kaoriImport = t.importDeclaration(imports, t.stringLiteral(KAORI_PACKAGE_NAME));
                        path.unshiftContainer("body", kaoriImport);
                    }
                },
            },
            ImportDeclaration(path, state) {
                // Check if component or html are already imported from kaori
                if (t.isStringLiteral(path.node.source) &&
                    (path.node.source.value === KAORI_PACKAGE_NAME ||
                        path.node.source.value.includes("kaori"))) {
                    path.node.specifiers.forEach((spec) => {
                        if (t.isImportSpecifier(spec) && t.isIdentifier(spec.imported)) {
                            if (spec.imported.name === "component") {
                                state.hasComponentImport = true;
                                state.componentFunctionName = spec.local.name;
                            }
                            if (spec.imported.name === "html") {
                                state.hasHtmlImport = true;
                                state.htmlTagName = spec.local.name;
                            }
                        }
                    });
                }
            },
            JSXElement(path, state) {
                const { node } = path;
                const tagName = getJSXElementName(node.openingElement.name);
                // Check if this is a component (starts with capital letter)
                if (isComponent(tagName)) {
                    state.needsComponentImport = true;
                    path.replaceWith(createComponentCall(node, tagName, state));
                }
                else {
                    state.needsHtmlImport = true;
                    path.replaceWith(createTemplateElement(node, state));
                }
            },
            JSXFragment(path, state) {
                // Handle fragments by wrapping content in template literal
                state.needsHtmlImport = true;
                const children = processJSXChildren(path.node.children, state);
                if (children.length === 0) {
                    // Empty fragment
                    const templateLiteral = t.templateLiteral([t.templateElement({ raw: "", cooked: "" }, true)], []);
                    path.replaceWith(t.taggedTemplateExpression(t.identifier(state.htmlTagName), templateLiteral));
                }
                else if (children.length === 1) {
                    // Single child - just wrap in html``
                    const templateLiteral = t.templateLiteral([
                        t.templateElement({ raw: "", cooked: "" }, false),
                        t.templateElement({ raw: "", cooked: "" }, true),
                    ], [children[0]]);
                    path.replaceWith(t.taggedTemplateExpression(t.identifier(state.htmlTagName), templateLiteral));
                }
                else {
                    // Multiple children - create array expression
                    path.replaceWith(t.arrayExpression(children));
                }
            },
        },
    };
}
function getJSXElementName(name) {
    if (t.isJSXIdentifier(name)) {
        return name.name;
    }
    if (t.isJSXMemberExpression(name)) {
        return `${getJSXElementName(name.object)}.${name.property.name}`;
    }
    if (t.isJSXNamespacedName(name)) {
        return `${name.namespace.name}:${name.name.name}`;
    }
    return "";
}
function isComponent(tagName) {
    return /^[A-Z]/.test(tagName);
}
function isSelfClosingTag(tagName) {
    const selfClosingTags = [
        "area",
        "base",
        "br",
        "col",
        "embed",
        "hr",
        "img",
        "input",
        "link",
        "meta",
        "param",
        "source",
        "track",
        "wbr",
    ];
    return selfClosingTags.includes(tagName.toLowerCase());
}
function normalizeAttributeName(attrName) {
    // Convert className to class
    if (attrName === "className")
        return "class";
    return attrName;
}
function createComponentCall(element, componentName, state) {
    const props = createPropsObject(element.openingElement.attributes);
    const children = processJSXChildren(element.children, state);
    // Add children to props if they exist
    if (children.length > 0) {
        const childrenValue = children.length === 1 ? children[0] : t.arrayExpression(children);
        if (t.isObjectExpression(props)) {
            props.properties.push(t.objectProperty(t.identifier("children"), childrenValue));
        }
    }
    return t.callExpression(t.identifier(state.componentFunctionName), [
        t.identifier(componentName),
        props,
    ]);
}
function createPropsObject(attributes) {
    const properties = [];
    for (const attr of attributes) {
        if (t.isJSXSpreadAttribute(attr)) {
            // Handle spread attributes
            properties.push(t.spreadElement(attr.argument));
        }
        else if (t.isJSXAttribute(attr)) {
            const propName = getAttributeName(attr);
            const propValue = getAttributeValue(attr);
            if (propName && propValue) {
                // Check if value needs getter wrapping (function calls, member access)
                if (needsGetterWrapping(propValue)) {
                    const getter = t.objectMethod("get", t.identifier(propName), [], t.blockStatement([t.returnStatement(propValue)]));
                    properties.push(getter);
                }
                else {
                    properties.push(t.objectProperty(t.identifier(propName), propValue));
                }
            }
        }
    }
    return t.objectExpression(properties);
}
function getAttributeName(attr) {
    if (t.isJSXIdentifier(attr.name)) {
        return attr.name.name;
    }
    if (t.isJSXNamespacedName(attr.name)) {
        return `${attr.name.namespace.name}:${attr.name.name.name}`;
    }
    return null;
}
function getAttributeValue(attr) {
    if (!attr.value) {
        return t.booleanLiteral(true);
    }
    if (t.isStringLiteral(attr.value)) {
        return attr.value;
    }
    if (t.isJSXExpressionContainer(attr.value)) {
        return attr.value.expression;
    }
    return null;
}
function needsGetterWrapping(expression) {
    let hasDynamicAccess = false;
    // Traverse the expression tree to find member access or function calls
    function traverse(node) {
        if (hasDynamicAccess)
            return; // Early exit if already found
        if (t.isMemberExpression(node) || t.isCallExpression(node)) {
            hasDynamicAccess = true;
            return;
        }
        // Stop traversal at function boundaries (inline functions should not cause wrapping)
        if (t.isArrowFunctionExpression(node) || t.isFunctionExpression(node)) {
            hasDynamicAccess = true; // Functions themselves need wrapping
            return; // But don't traverse into their bodies
        }
        // Continue traversing child nodes
        for (const key in node) {
            const child = node[key];
            if (child && typeof child === "object") {
                if (Array.isArray(child)) {
                    for (const item of child) {
                        if (item && typeof item === "object" && item.type) {
                            traverse(item);
                        }
                    }
                }
                else if (child.type) {
                    traverse(child);
                }
            }
        }
    }
    traverse(expression);
    return hasDynamicAccess;
}
function createTemplateElement(element, state) {
    const tagName = getJSXElementName(element.openingElement.name);
    const attributes = element.openingElement.attributes;
    const isSelfClosing = isSelfClosingTag(tagName);
    // Build the template string parts and expressions
    const templateParts = [];
    const expressions = [];
    // Start with opening tag
    let currentPart = `<${tagName}`;
    // Process attributes
    for (const attr of attributes) {
        if (t.isJSXAttribute(attr)) {
            const attrName = getAttributeName(attr);
            const attrValue = getAttributeValue(attr);
            if (attrName && attrValue) {
                const normalizedAttrName = normalizeAttributeName(attrName);
                if (attrName.startsWith("on")) {
                    // Event handlers: onClick -> @click=${}
                    const eventName = attrName.slice(2).toLowerCase();
                    currentPart += ` @${eventName}=`;
                    templateParts.push(currentPart);
                    expressions.push(attrValue);
                    currentPart = "";
                }
                else if (attrName.includes(":")) {
                    const [namespace, name] = attrName.split(":");
                    if (namespace === "prop") {
                        // prop:value -> .value=${}
                        currentPart += ` .${name}=`;
                        templateParts.push(currentPart);
                        expressions.push(attrValue);
                        currentPart = "";
                    }
                    else if (namespace === "bool") {
                        // bool:name -> ?name=${}
                        currentPart += ` ?${name}=`;
                        templateParts.push(currentPart);
                        expressions.push(attrValue);
                        currentPart = "";
                    }
                    else {
                        // Regular namespaced attribute
                        currentPart += ` ${attrName}="`;
                        templateParts.push(currentPart);
                        expressions.push(attrValue);
                        currentPart = `"`;
                    }
                }
                else if (t.isStringLiteral(attrValue)) {
                    // Regular string attribute
                    currentPart += ` ${normalizedAttrName}="${attrValue.value}"`;
                }
                else {
                    // Dynamic attribute
                    currentPart += ` ${normalizedAttrName}=`;
                    templateParts.push(currentPart);
                    expressions.push(attrValue);
                    currentPart = "";
                }
            }
        }
    }
    // Close opening tag and handle children/closing
    if (isSelfClosing) {
        // Self-closing tag - just close with />
        currentPart += " />";
        templateParts.push(currentPart);
    }
    else {
        // Regular element - close opening tag and add children
        currentPart += ">";
        // Process children more efficiently - combine static content
        const childResults = processJSXChildrenForTemplate(element.children, state);
        for (const result of childResults) {
            if (result.type === "static") {
                currentPart += result.content;
            }
            else {
                // Dynamic content - need expression
                templateParts.push(currentPart);
                expressions.push(result.expression);
                currentPart = "";
            }
        }
        // Add closing tag
        currentPart += `</${tagName}>`;
        templateParts.push(currentPart);
    }
    // Create template elements - must be exactly one more than expressions
    const templateElements = templateParts.map((part, index) => t.templateElement({ raw: part, cooked: part }, index === templateParts.length - 1));
    const templateLiteral = t.templateLiteral(templateElements, expressions);
    return t.taggedTemplateExpression(t.identifier(state.htmlTagName), templateLiteral);
}
function processJSXChildrenForTemplate(children, state) {
    const results = [];
    for (const child of children) {
        if (t.isJSXText(child)) {
            // Include text directly in the template (preserve whitespace for inline text)
            const text = child.value;
            if (text.trim()) {
                // Has meaningful content, add as static text
                results.push({ type: "static", content: text });
            }
            else if (text.includes("\n")) {
                // Just whitespace with newlines, skip it
                continue;
            }
            else {
                // Inline whitespace, preserve it
                results.push({ type: "static", content: text });
            }
        }
        else if (t.isJSXExpressionContainer(child)) {
            if (t.isExpression(child.expression)) {
                results.push({ type: "dynamic", expression: child.expression });
            }
        }
        else if (t.isJSXElement(child)) {
            const tagName = getJSXElementName(child.openingElement.name);
            if (isComponent(tagName)) {
                state.needsComponentImport = true;
                results.push({
                    type: "dynamic",
                    expression: createComponentCall(child, tagName, state),
                });
            }
            else {
                // For HTML elements, check if they have only simple attributes and text content
                if (canInlineAsStaticHTML(child)) {
                    results.push({
                        type: "static",
                        content: convertElementToStaticHTML(child),
                    });
                }
                else {
                    // Instead of creating nested html templates, inline the element structure
                    const elementResult = inlineElementInTemplate(child, state);
                    results.push(...elementResult);
                }
            }
        }
        else if (t.isJSXFragment(child)) {
            const fragmentResults = processJSXChildrenForTemplate(child.children, state);
            results.push(...fragmentResults);
        }
    }
    return results;
}
function canInlineAsStaticHTML(element) {
    // Can inline if element has no dynamic attributes and all children are static
    const hasOnlyStaticAttributes = element.openingElement.attributes.every((attr) => {
        if (t.isJSXAttribute(attr)) {
            const attrValue = getAttributeValue(attr);
            return (!attrValue ||
                t.isStringLiteral(attrValue) ||
                t.isBooleanLiteral(attrValue));
        }
        return false; // spread attributes are dynamic
    });
    if (!hasOnlyStaticAttributes)
        return false;
    // Check if all children are static text or static HTML elements
    return element.children.every((child) => {
        if (t.isJSXText(child))
            return true;
        if (t.isJSXElement(child)) {
            const tagName = getJSXElementName(child.openingElement.name);
            return !isComponent(tagName) && canInlineAsStaticHTML(child);
        }
        return false; // expressions or fragments are dynamic
    });
}
function convertElementToStaticHTML(element) {
    const tagName = getJSXElementName(element.openingElement.name);
    const isSelfClosing = isSelfClosingTag(tagName);
    let html = `<${tagName}`;
    // Add attributes
    for (const attr of element.openingElement.attributes) {
        if (t.isJSXAttribute(attr)) {
            const attrName = getAttributeName(attr);
            const attrValue = getAttributeValue(attr);
            if (attrName && attrValue) {
                const normalizedAttrName = normalizeAttributeName(attrName);
                if (t.isStringLiteral(attrValue)) {
                    html += ` ${normalizedAttrName}="${attrValue.value}"`;
                }
                else if (t.isBooleanLiteral(attrValue) && attrValue.value) {
                    html += ` ${normalizedAttrName}`;
                }
            }
        }
    }
    if (isSelfClosing) {
        html += " />";
    }
    else {
        html += ">";
        // Add children
        for (const child of element.children) {
            if (t.isJSXText(child)) {
                html += child.value;
            }
            else if (t.isJSXElement(child)) {
                html += convertElementToStaticHTML(child);
            }
        }
        html += `</${tagName}>`;
    }
    return html;
}
function inlineElementInTemplate(element, state) {
    const tagName = getJSXElementName(element.openingElement.name);
    const attributes = element.openingElement.attributes;
    const isSelfClosing = isSelfClosingTag(tagName);
    const results = [];
    // Start with opening tag
    let currentPart = `<${tagName}`;
    // Process attributes
    for (const attr of attributes) {
        if (t.isJSXAttribute(attr)) {
            const attrName = getAttributeName(attr);
            const attrValue = getAttributeValue(attr);
            if (attrName && attrValue) {
                const normalizedAttrName = normalizeAttributeName(attrName);
                if (attrName.startsWith("on")) {
                    // Event handlers: onClick -> @click=${}
                    const eventName = attrName.slice(2).toLowerCase();
                    currentPart += ` @${eventName}=`;
                    results.push({ type: "static", content: currentPart });
                    results.push({ type: "dynamic", expression: attrValue });
                    currentPart = "";
                }
                else if (attrName.includes(":")) {
                    const [namespace, name] = attrName.split(":");
                    if (namespace === "prop") {
                        // prop:value -> .value=${}
                        currentPart += ` .${name}=`;
                        results.push({ type: "static", content: currentPart });
                        results.push({ type: "dynamic", expression: attrValue });
                        currentPart = "";
                    }
                    else if (namespace === "bool") {
                        // bool:name -> ?name=${}
                        currentPart += ` ?${name}=`;
                        results.push({ type: "static", content: currentPart });
                        results.push({ type: "dynamic", expression: attrValue });
                        currentPart = "";
                    }
                    else {
                        // Regular namespaced attribute
                        currentPart += ` ${attrName}="`;
                        results.push({ type: "static", content: currentPart });
                        results.push({ type: "dynamic", expression: attrValue });
                        currentPart = `"`;
                    }
                }
                else if (t.isStringLiteral(attrValue)) {
                    // Regular string attribute
                    currentPart += ` ${normalizedAttrName}="${attrValue.value}"`;
                }
                else {
                    // Dynamic attribute
                    currentPart += ` ${normalizedAttrName}=`;
                    results.push({ type: "static", content: currentPart });
                    results.push({ type: "dynamic", expression: attrValue });
                    currentPart = "";
                }
            }
        }
    }
    // Close opening tag and handle children/closing
    if (isSelfClosing) {
        // Self-closing tag - just close with />
        currentPart += " />";
        results.push({ type: "static", content: currentPart });
    }
    else {
        // Regular element - close opening tag and add children
        currentPart += ">";
        // Process children
        const childResults = processJSXChildrenForTemplate(element.children, state);
        if (childResults.length === 0) {
            // No children, just close the tag
            currentPart += `</${tagName}>`;
            results.push({ type: "static", content: currentPart });
        }
        else {
            // Add opening tag part
            if (currentPart) {
                results.push({ type: "static", content: currentPart });
            }
            // Add children
            results.push(...childResults);
            // Add closing tag
            results.push({ type: "static", content: `</${tagName}>` });
        }
    }
    return results;
}
function processJSXChildren(children, state) {
    const processedChildren = [];
    for (const child of children) {
        if (t.isJSXText(child)) {
            // Only include text if it has non-whitespace content
            const text = child.value.trim();
            if (text) {
                processedChildren.push(t.stringLiteral(text));
            }
        }
        else if (t.isJSXExpressionContainer(child)) {
            if (t.isExpression(child.expression)) {
                processedChildren.push(child.expression);
            }
        }
        else if (t.isJSXElement(child)) {
            const tagName = getJSXElementName(child.openingElement.name);
            if (isComponent(tagName)) {
                state.needsComponentImport = true;
                processedChildren.push(createComponentCall(child, tagName, state));
            }
            else {
                state.needsHtmlImport = true;
                processedChildren.push(createTemplateElement(child, state));
            }
        }
        else if (t.isJSXFragment(child)) {
            const fragmentChildren = processJSXChildren(child.children, state);
            processedChildren.push(...fragmentChildren);
        }
    }
    return processedChildren;
}
export default jsxToLitHtmlPlugin;
//# sourceMappingURL=babel-plugin.js.map