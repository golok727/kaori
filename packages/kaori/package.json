{
  "name": "kaori.js",
  "version": "0.0.4",
  "description": "A fragrant TS framework",
  "type": "module",
  "exports": {
    ".": "./src/index.ts",
    "./jsx-runtime": "./src/jsx/types.ts",
    "./jsx-dev-runtime": "./src/jsx/types.ts"
  },
  "publishConfig": {
    "exports": {
      ".": "./dist/index.js",
      "./jsx-runtime": "./dist/jsx/types.d.ts",
      "./jsx-dev-runtime": "./dist/jsx/types.d.ts"
    }
  },
  "files": [
    "dist",
    "src",
    "package.json"
  ],
  "scripts": {
    "build": "tsc"
  },
  "keywords": [
    "framework",
    "web",
    "frontend",
    "ui",
    "typescript",
    "kaori",
    "reactive",
    "signals",
    "lit-html"
  ],
  "author": "golok",
  "packageManager": "pnpm@10.12.4",
  "dependencies": {
    "@preact/signals-core": "^1.12.1",
    "lit-html": "^3.3.1"
  }
}
