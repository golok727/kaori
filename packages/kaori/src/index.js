import { html } from "lit-html";
import { AsyncDirective, directive } from "lit-html/async-directive.js";
import { effect as syncEffect, untracked } from "@preact/signals-core";
import { invariant } from "./utils.js";
const logger = {
    log: (...args) => {
        console.log("[Kaori:DEV]", ...args);
    },
};
// todo move to a runtime file
/*
 Globals
 -----------------------------------------------------------------------
*/
let active_bloom = null;
let is_micro_task_queued = false;
let queued_microtasks = [];
let queued_updates = [];
let dirty_blooms = new WeakSet();
/*--------------------------------------------------------------- */
function queue_microtask(fn) {
    if (!is_micro_task_queued) {
        is_micro_task_queued = true;
        queueMicrotask(flush_microtasks);
    }
    if (fn) {
        queued_microtasks.push(fn);
    }
}
function flush_microtasks() {
    is_micro_task_queued = false;
    if (queued_microtasks.length > 0) {
        let microtasks = queued_microtasks;
        queued_microtasks = [];
        for (let i = 0; i < microtasks.length; i++) {
            microtasks[i]();
        }
    }
    const needs_updates = queued_updates;
    queued_updates = [];
    for (let i = 0; i < needs_updates.length; i++) {
        const [bloom, fn] = needs_updates[i];
        dirty_blooms.delete(bloom);
        fn();
    }
}
function schedule_update(bloom, fn) {
    if (dirty_blooms.has(bloom)) {
        return;
    }
    dirty_blooms.add(bloom);
    if (!is_micro_task_queued) {
        is_micro_task_queued = true;
        queueMicrotask(flush_microtasks);
    }
    queued_updates.push([bloom, fn]);
}
function on_mount(fn) {
    invariant(active_bloom !== null, "onMount() should be called during component setup (not in render functions or effects)");
    queue_microtask(fn);
}
function on_cleanup(fn) {
    const bloom = active_bloom;
    invariant(bloom !== null, "onCleanup() should be called during component setup (not in render functions or effects)");
    bloom.__disposables.add(fn);
}
function bloom_effect(fn, bloom = active_bloom) {
    invariant(bloom !== null, "effect() should be called during component setup (not in render functions or effects)");
    const dispose = untracked(() => syncEffect(fn));
    bloom.__disposables.add(dispose);
}
function get_bloom() {
    invariant(active_bloom !== null, "getBloom() can only be called during component setup (not in render functions or effects)");
    return active_bloom;
}
function dispose_bloom(bloom) {
    logger.log("Disposing bloom and running cleanup functions");
    bloom.__disposables.forEach((dispose) => {
        try {
            dispose();
        }
        catch (e) {
            console.error("Error during cleanup:", e);
        }
    });
    bloom.__disposables.clear();
}
class ComponentDirective extends AsyncDirective {
    constructor() {
        super(...arguments);
        this._rawTemplate = null;
        this.bloom = null;
        this.hasInitialized = false;
        this._cachedTemplate = null;
    }
    get $_getTemplate() {
        return typeof this._rawTemplate === "function"
            ? this._rawTemplate()
            : this._rawTemplate;
    }
    render(C, props) {
        const componentName = C.name || "Anonymous";
        if (!this.hasInitialized) {
            logger.log("Rendering component ", componentName, props);
            // Create the bloom object with update method
            this.bloom = {
                update: () => {
                    this._cachedTemplate = this.$_getTemplate;
                    if (this.bloom) {
                        schedule_update(this.bloom, () => {
                            if (this.isConnected) {
                                this.setValue(this._cachedTemplate);
                            }
                        });
                    }
                },
                __disposables: new Set(),
            };
            // Set current bloom context and call component
            const prevBloom = active_bloom;
            active_bloom = this.bloom;
            const result = untracked(() => C(props));
            active_bloom = prevBloom;
            this._rawTemplate = result;
            if (typeof this._rawTemplate === "function") {
                // reactive return
                bloom_effect(() => {
                    logger.log("Effect, ", componentName);
                    this.bloom?.update();
                }, this.bloom);
            }
            this._cachedTemplate = this.$_getTemplate;
            this.hasInitialized = true;
        }
        return this._cachedTemplate;
    }
    disconnected() {
        // Clean up the effect when component unmounts
        if (this.bloom) {
            dispose_bloom(this.bloom);
        }
    }
    reconnected() {
        // we need to do something here :w
        //
    }
}
export const component = directive(ComponentDirective);
export { For, Show } from "./helpers";
export { get_bloom as getBloom, bloom_effect as effect, on_mount as onMount, on_cleanup as onCleanup, html, };
export { render } from "lit-html";
export * from "./types";
export { signal, batch, computed, effect as syncEffect, untracked, } from "@preact/signals-core";
export { nothing } from "lit-html";
export { when } from "lit-html/directives/when.js";
export { choose } from "lit-html/directives/choose.js";
export { repeat } from "lit-html/directives/repeat.js";
export { map } from "lit-html/directives/map.js";
export { ref, createRef } from "lit-html/directives/ref.js";
//# sourceMappingURL=index.js.map