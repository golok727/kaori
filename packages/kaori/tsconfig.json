{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "moduleResolution": "node",
    "tsBuildInfoFile": "./dist/tsconfig.tsbuildinfo"
  },
  "include": ["src", "src/jsx"]
}
