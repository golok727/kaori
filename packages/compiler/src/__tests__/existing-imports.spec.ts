import { describe, it, expect } from "vitest";
import babel from "@babel/core";
import { KaoriCompiler } from "../babel-plugin.js";

async function transformJSX(jsxCode: string) {
  const result = await babel.transformAsync(jsxCode, {
    plugins: [["@babel/plugin-syntax-jsx"], [KaoriCompiler]],
    parserOpts: {
      plugins: ["jsx", "typescript"],
    },
  });
  return result?.code || "";
}

describe("JSX transformation with existing imports", () => {
  it("should handle existing kaori.js imports", async () => {
    const input = `
import {component, computed, html, render, signal} from "kaori.js";

function Kaori() {
    let count = 0;
    return () => html\`<h1 class="text-lg font-bold"> Without signals count \${count}</h1>\`;
}

function App() {
    return (
        <div>
            <h1>My App</h1>
            <Kaori />
            <p>Some content</p>
        </div>
    );
}`;

    const output = await transformJSX(input);
    expect(output).toMatchSnapshot();
  });

  it("should handle existing kaori imports from path", async () => {
    const input = `
import {component, computed, html, render, signal, For, getBloom, onCleanup, Show, nothing, onMount} from "kaori.js";

function KaoriThing() {
    let count = 0;
    return () => html\`<h1 class="text-lg font-bold"> Without signals count \${count}</h1>\`;
}

function App() {
    return (
        <div>
            <h1>My App</h1>
            <KaoriThing />
            <p>Some content</p>
        </div>
    );
}`;

    const output = await transformJSX(input);
    expect(output).toMatchSnapshot();
  });

  it("should handle existing html only import", async () => {
    const input = `
import {html, render} from "kaori.js";

function App() {
    return (
        <div>
            <h1>My App</h1>
            <MyComponent />
        </div>
    );
}`;

    const output = await transformJSX(input);
    expect(output).toMatchSnapshot();
  });

  it("should handle existing component only import", async () => {
    const input = `
import {component, render} from "kaori.js";

function App() {
    return (
        <div>
            <h1>My App</h1>
            <MyComponent />
        </div>
    );
}`;

    const output = await transformJSX(input);
    expect(output).toMatchSnapshot();
  });

  it("should handle aliased imports", async () => {
    const input = `
import {component as comp, html as h} from "kaori.js";

function App() {
    return (
        <div>
            <h1>My App</h1>
            <MyComponent />
        </div>
    );
}`;

    const output = await transformJSX(input);
    expect(output).toMatchSnapshot();
  });
});
