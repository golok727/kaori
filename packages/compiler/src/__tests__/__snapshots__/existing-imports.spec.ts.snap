// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JSX transformation with existing imports > should handle aliased imports 1`] = `
"import { component as comp, html as h } from "kaori.js";
function App() {
  return h\`<div><h1>My App</h1>\${comp(MyComponent, {})}</div>\`;
}"
`;

exports[`JSX transformation with existing imports > should handle existing component only import 1`] = `
"import { html } from "kaori.js";
import { component, render } from "kaori.js";
function App() {
  return html\`<div><h1>My App</h1>\${component(MyComponent, {})}</div>\`;
}"
`;

exports[`JSX transformation with existing imports > should handle existing html only import 1`] = `
"import { component } from "kaori.js";
import { html, render } from "kaori.js";
function App() {
  return html\`<div><h1>My App</h1>\${component(MyComponent, {})}</div>\`;
}"
`;

exports[`JSX transformation with existing imports > should handle existing kaori imports from path 1`] = `
"import { component, computed, html, render, signal, For, getBloom, onCleanup, Show, nothing, onMount } from "kaori.js";
function KaoriThing() {
  let count = 0;
  return () => html\`<h1 class="text-lg font-bold"> Without signals count \${count}</h1>\`;
}
function App() {
  return html\`<div><h1>My App</h1>\${component(KaoriThing, {})}<p>Some content</p></div>\`;
}"
`;

exports[`JSX transformation with existing imports > should handle existing kaori.js imports 1`] = `
"import { component, computed, html, render, signal } from "kaori.js";
function Kaori() {
  let count = 0;
  return () => html\`<h1 class="text-lg font-bold"> Without signals count \${count}</h1>\`;
}
function App() {
  return html\`<div><h1>My App</h1>\${component(Kaori, {})}<p>Some content</p></div>\`;
}"
`;
